tools/respecDocWriter.js

```diff
@@ -116,6 +125,67 @@ async function fetchAndWrite(
   }
 }
 
+/**
+ * Replace the ReSpec script in document with the locally installed one. This is
+ * useful in CI env or when you want to pin the ReSpec version.
+ *
+ * @assumption The ReSpec script being used in the document is hosted on either
+ * w3.org or w3c.github.io. If this assumption doesn't hold true (interception
+ * fails), this function will timeout.
+ *
+ * The following ReSpec URLs are supported:
+ * https://www.w3.org/Tools/respec/${profile}
+ * https://w3c.github.io/respec/builds/${profile}.js
+ * file:///home/path-to-respec/builds/${profile}.js
+ * http://localhost:PORT/builds/${profile}.js
+ * https://example.com/builds/${profile}.js
+ *
+ * @param {import("puppeteer").Page} page
+ */
+async function useLocalReSpec(page) {
+  await page.setRequestInterception(true);
+
+  page.on("request", async function requestInterceptor(request) {
+    if (!isRespecScript(request)) {
+      await request.continue();
+      return;
+    }
+
+    const url = new URL(request.url());
+    const respecProfileRegex = /\/(respec-[\w-]+)(?:\.js)?$/;
+    const profile = url.pathname.match(respecProfileRegex)[1];
+    const localPath = path.join(__dirname, "..", "builds", `${profile}.js`);
+    console.log(colors.info(`Intercepted ${url} to respond with ${localPath}`));
+    await request.respond({
+      contentType: "text/javascript; charset=utf-8",
+      body: await readFile(localPath),
+    });
+    // Workaround for https://github.com/puppeteer/puppeteer/issues/4208
+    page.removeListener("request", requestInterceptor);
+    await page.setRequestInterception(false);
+  });
+}
+
+/** @param {import("puppeteer").Request} req */
+function isRespecScript(req) {
+  if (req.method() !== "GET" || req.resourceType() !== "script") {
+    return false;
+  }
+
+  const { host, pathname: path } = new URL(req.url());
+  switch (host) {
+    case "www.w3.org":
+      return (
+        path.startsWith("/Tools/respec/") && !path.includes("respec-highlight")
+      );
+    case "w3c.github.io":
+      return path.startsWith("/respec/builds/");
+    default:
+      // localhost, file://, and everything else
+      return /\/builds\/respec-[\w-]+\.js$/.test(path);
```

Comment:
Why not just `respec-[\w-]+\.js`?

Reply:
Could also match respec-worker, respec-highlight

Reply:
Yes, I mean why `builds/`?

Reply:
We're only replacing `respec-[profile]` here, which is in `builds/`. Let me try if we can do without `/builds/`

Reply:
Adding `/builds/` reduces risk of false-positive though. A mismatch here means crash. Say, some adds their config as `respec-config.js` (haven't seen cases, but possible). `builds/respec-config.js` is less likely.

Reply:
Is this story possible?: We catch `respec-config.js` or whatever we don't support, but it's replacement is not available, so we just passthrough in this case.

Reply:
Due to https://github.com/puppeteer/puppeteer/issues/4208, we need to disable request interception as soon as possible (otherrwise worker script crashes). So, I'm hesitant to try to match more scripts. Ideally, I would like to match respec-worker and respec-highlight also (and probably support an entirely offline experience)

Reply:
Hmm. Will migration to playwright help? 🤔

Reply:
So I played with playwright a bit. Although I didn't check if this problem exists there, playwright has its own problems. 

Playwright not supporting using existing browser binaries is a major blocker in my opinion. One can [install all/some binaries in a folder and then re-use them](https://github.com/microsoft/playwright/blob/master/docs/installation.md), but it doesn't let me reuse the pre-installed browsers. Puppeteer allows it, and we do use it in github-actions (I also use it locally).

Reply:
It may not be a blocker though:

1. The download already happens for Firefox run, and we already are caching the dependencies via `actions/cache`.
1. Looking at https://github.com/microsoft/playwright/issues/2623#issuecomment-646051077, it seems there is `executablePath` to use a custom path, although I didn't find a way to skip the download.

Reply:
Download from cache is kinda slow though. Given we already have a browser installed, it'll definitely feel slower to not use it.

PLAYWRIGHT_SKIP_BROWSER_DOWNLOAD=1 should let us skip download though.

Nice find on passing executable path. Can set it with env variables.

I'll play more with it, if we do want to go with playwright. Though, I'm okay with puppeteer also.

Reply:
It's https://github.com/microsoft/playwright/issues/5952 vs https://github.com/puppeteer/puppeteer/issues/4208 now.

Reply:
@saschanaz I think intercepting only `builds/respec-{profile}` should be good enough for now?

Reply:
Yeah, it seem it won't be fixed anytime soon.

Reply:
@saschanaz importScripts from workers should now be fixed on vNext/master of playwright. When v1.11.0 comes out, feel free to give it a try and raise an issue if you encounter problems. 😄 

NB: This fix should simply prevent the importScripts from hanging; I'm not sure of the general story for request interception of SW traffic.